agent/command.go

```diff
@@ -146,15 +146,22 @@ func (a *Agent) runCommandOrFunc(ctx context.Context, tc *taskContext, commandIn
 // single sub-command within a function.
 func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.LoggerProducer, commandInfo model.PluginCommandConf,
 	cmd command.Command, displayName string, options runCommandsOptions) error {
-
+	prevDef := map[string]string{}
 	for key, val := range commandInfo.Vars {
+		prevVal := tc.taskConfig.Expansions.Get(key)
+		prevDef[key] = prevVal
+
 		var newVal string
 		newVal, err := tc.taskConfig.Expansions.ExpandString(val)
 		if err != nil {
 			return errors.Wrapf(err, "expanding '%s'", val)
 		}
 		tc.taskConfig.Expansions.Put(key, newVal)
 	}
+	// unset the function vars once the function finished running
```

Comment:
I don't think we need this comment since it's just stating what the code is doing. However, if you want to explain the way, you can change the comment to say you want to ensure function vars do not persist in the expansions after the function is over (inside the defer block).